Frontend charm integration tests fix with a self hosted runner

mz2 opened the pull request:
Integration tests for the frontend charm now actually work also on GitHub!

The workflow looks to be stable, with a dozen passes witnessed today.

It does however take ~10mins to run and I figured we would therefore not start running it on every commit to every branch (6+ mins of the time is taken by the microk8s setup -- we can drop this in a big way with custom images used for the CI runs when that becomes available). It runs on:
- all branches once a PR review has been posted for the branch for which there is a PR open
- all branches with a PR which touch resources under `./frontend/charm` (the integration tests can capture failures also in resources outside that directory, hence the rule above also).
- on the main branch
- on version tagged commits
- manually via workflow dispatch

The workflow was refactored to reuse two now introduced reusable composite actions in https://github.com/canonical/certification-github-workflows (the [first PR to that repo](https://github.com/canonical/certification-github-workflows/pull/2) should be reviewed and merged first before this goes in).

**NOTE!** When squashing this into main, please edit the commit message heavily down in volume since this contains a few months worth of desperate debugging attempts of these integration tests that continued to work locally just fine but failed to be executed specifically on GitHub up until now. I have simplified the version history with some squashing but since there are several merge commits there I didn't want to do anything more drastic than that.

mz2 committed:
Define the dependencies more exactly to get the integration tests running.

mz2 committed:
juju 3.1/stable -> 3.0/stable

mz2 committed:
Install strictly confined versions of juju and microk8s.

mz2 committed:
Use microk8s 1.27-strict/stable

mz2 committed:
Adds package repo access to the integration test job

mz2 committed:
Add the containerd template modification

mz2 committed:
Adds some missing sudoing

mz2 committed:
From >> to tee -a

mz2 committed:
Drop the docker login

mz2 committed:
Oops, escaping quotes did not work quite right there.

mz2 committed:
Adjust the integration test scenario, expecting rather than raising on blocked status (more scenarios needed to actually relate the frontend to the backend).

mz2 committed:
Drops the debugging intended branch from the list of branches to trigger the integration tests with.

mz2 committed:
Change to maintenance status instead for clarity, since configuration action is required.

mz2 committed:
Adds the "integration-test-fix" branch into the rule still

mz2 committed:
Adds a 503 response / maintenance status to the frontend application state where relation is not yet connected.

Simplified the charm logic, no _stored state needed (fetching the hostname and port from the relation's application data bag).

mz2 committed:
Removs the redundant import StoredState

mz2 committed:
Add a `microk8s status --wait-ready`

mz2 committed:
Adds missing sudo

nadzyah committed:
Merge branch 'main' into integration-test-fix

nadzyah committed:
Merge branch 'main' into integration-test-fix

nadzyah committed:
Merge branch 'main' into integration-test-fix

mz2 committed:
Let's try with the microk8s version + libpyjuju < 3.1

Let's try with the microk8s version and libpyjuju < 3.1 that seems to work for rout53-acme-operator where the integration test pipeline is pulled from

mz2 committed:
Merge branch 'integration-test-fix' of github.com:canonical/test_observer into integration-test-fix

mz2 committed:
Back to 5.13/stable for lxd

mz2 committed:
2.9/stable -> 3.0/stable

mz2 committed:
Trying with the non-strictly confined microk8s

mz2 committed:
Adds "publish-charm" job

mz2 committed:
Adds promotion action

mz2 committed:
Swap to self-hosted runners

mz2 committed:
LXD 5.13 => 5.19

mz2 committed:
Run also the integration tests

mz2 committed:
Move publishing charm to a separate workflow

mz2 committed:
juju 3.0 => juju 3.2

mz2 committed:
juju 3.0 -> 3.2

mz2 committed:
Adds a now required trailing comma

mz2 committed:
microk8s 1.27 => 1.28

mz2 committed:
Use mz2 fork of the action operator

mz2 committed:
Juju classic confinement = off too

mz2 committed:
Pulls in origin/main

plars reviewed (changes requested):
A few questions below

mz2 commented on .github/workflows/frontend_charm_integration_tests.yml:
```diff
@@ -1,24 +1,52 @@
+          channel: 5.19/stable
       - name: Setup operator environment
-        uses: charmed-kubernetes/actions-operator@main
+        uses: mz2/actions-operator@4fcb55b99e2a0f69a5910870d3639d203aeb87aa
```
Because otherwise I can't force strictly confined versions of juju and microk8s to be used.

mz2 replied:
I have posted a PR with my changes, indeed should be able to merge back.

For clarity this branch when in a mergeable state will be the first point at which these charm integration tests actually behave.

mz2 replied:
Plot twist: turns out my tweaks here were unnecessary because snap ignores `--classic` for a strictly confined snap: https://github.com/charmed-kubernetes/actions-operator/pull/57 --> I have changed this back to the way it was before.

mz2 commented on frontend/charm/metadata.yaml:
```diff
@@ -6,7 +6,7 @@ description: |
 
 assumes:
-  - juju >= 2.9
+  - juju >= 3.1
```
(I doubt it'll be a problem since with this present at 2.9, juju 3.1 and 3.2 have run the application successfully in local dev envs)

mz2 replied:
I have changed this back to 2.9 and it all works (well, it builds and integration and unit tests pass anyway) still.

mz2 committed:
Introduces a working integration testing workflow for the frontend charm

The workflow depends on self hosted runner and composite actions introduced in https://github.com/canonical/certification-github-workflows

mz2 commented on .github/workflows/publish_charms.yml:
```diff
@@ -0,0 +1,32 @@
+        uses: actions/checkout@v2
+      - name: Install charmcraft
+        run: sudo snap install charmcraft --classic
+      - name: Fetch Tested Charm
```
```suggestion
      - name: Fetch the tested charm
```

mz2 committed:
Update .github/workflows/publish_charms.yml

mz2 committed:
Merges in origin/main

mz2 committed:
Merge branch 'integration-test-fix-with-hosted-runners' of github.com:canonical/test_observer into integration-test-fix-with-hosted-runners

mz2 committed:
Remove publish_charms.yml that has been long since been separately added

mz2 committed:
Removes release-charms.yml that has long been superseded

mz2 committed:
Removes also frontend_charm_promote.yml that is not really in scope of these changes

mz2 committed:
From 3.2 requirement => 2.9 requirement

mz2 commented on frontend/charm/requirements.txt:
```diff
@@ -1 +1,2 @@
-ops >= 2.2.0
+ops >= 2.8.0
+macaroonbakery==1.3.2
```
Added as a workaround to https://github.com/canonical/test_observer/actions/runs/7142353669/job/19451476954 as instructed at https://chat.canonical.com/canonical/pl/9xzxm61rgtry5d7sfx9xz7ryyc

omar-selo replied:
can you add a comment to remind us why this is needed?

mz2 committed:
Removes the stray reformatting of frontend charm.py

mz2 commented on frontend/charm/tox.ini:
```diff
@@ -66,7 +66,7 @@ commands =
 deps =
     pytest
-    juju
+    juju<=3.2
```
This is intentionally `<= 3.2` as opposed to for example `>= 2.9` as is set in the charm metadata: the integration test environment needs to be (for reasons of various version incompatibilities) 3.x.

mz2 commented on .github/workflows/frontend_charm_integration_tests.yml:
```diff
@@ -1,45 +1,26 @@
-          provider: microk8s
+        uses: actions/checkout@v4
+      - name: Set up microk8s
+        uses: canonical/certification-github-workflows/.github/actions/microk8s-setup@microk8s-setup-action
```
The setup steps turned out to be more extensive on self hosted runners, and since this is all repetitive boilerplate, moved to a reusable composite action. Same story below re: the `archive-charm-testing-artifacts`

mz2 committed:
Remove another example of accidental reformatting

mz2 committed:
Run the integration test workflow manually, on main branch and when PR review has been submitted

mz2 committed:
Run frontend integration tests also on branches where there are changes to resources under frontend/charm

mz2 committed:
Removes a stray change to frontend_charm_analysis.yml

mz2 commented on terraform/test-observer.tf:
```diff
@@ -1,7 +1,7 @@
 terraform {
   required_providers {
     juju = {
-      version = "~> 0.7.0"
```
The changes below showed up as necessary when the provider was updated: the `series` (and `base`) are nonsensical for k8s charms => removed.

mz2 replied:
This change is not strictly speaking necessary for the charm integration tests (though it was my intention to also test out in succession also that the terraform plan continues to be able to spin a new environment up that stabilises in an all-ready state).

omar-selo replied:
can we do this on PS5 though? as in can we use tf provider for juju version `0.10.0`?

mz2 replied:
I think this should indeed Just Work, in that the next time you deploy using this plan, it will upgrade the provider to the latest version that is described here. Deploying to staging will confirm that safely, I'd say.

omar-selo commented on .github/workflows/publish_api.yml:
```diff
@@ -8,7 +8,7 @@ env:
 jobs:
   build-and-push-image:
-    runs-on: ubuntu-latest
+    runs-on: [self-hosted, linux, xlarge]
```
this change will require setting up proxy so that images can be pushed to `ghcr`. Unless self hosted runners have been updated and don't require this anymore?

mz2 replied:
Indeed no longer needed. :partying_face:

mz2 replied:
... however, both here and in the frontend image publishing workflow it is necessary to still configure a mirror for docker.io such as not to hit rate limits. Done that in a follow-up commit now.

omar-selo commented on .github/workflows/publish_frontend.yml:
```diff
@@ -8,7 +8,7 @@ env:
 jobs:
   build-and-push-frontend-image:
-    runs-on: ubuntu-latest
+    runs-on: [self-hosted, linux]
```
Same here

mz2 replied:
Proxying no longer needed to access ghcr.io, but docker.io rate limits could still be hit --> mirror configuration added.

omar-selo reviewed (commented):
It's amazing that you got it to work! finally this has been annoying for a long time

omar-selo commented on .github/workflows/frontend_charm_integration_tests.yml:
```diff
@@ -1,45 +1,34 @@
-name: Run charm tests
+name: Run charm integration tests
 on:
+  pull_request_review:
+    types: [submitted]
```
perhaps we don't need this as it triggers whenever someone reviews the PR

mz2 replied:
Turned out I was missing an `if` condition as per https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#running-a-workflow-when-a-pull-request-is-approved (and as implemented in the metabox tests from github.com/canonical/checkbox where I copied this from)

mz2 committed:
An explanatory comment re: pinning macaroonbakery to 1.3.2

mz2 committed:
Adds docker.io mirroring

mz2 committed:
Fix the trigger so the workflow triggers on approved PRs

mz2 committed:
Reference the setup steps from main

mz2 commented:
Superseded by https://github.com/canonical/test_observer/pull/80
